### ISSUE TITLE
Intellisense/autocompletion not working for extenral crates with optional features

### ISSUE THREAD

**Issue by spamwax** [rust-lang/rls-vscode#637]
Although this may seem like a duplicate, I believe it's slightly different than other issues.

## TLDR
Main issue is that Intellisense works for external crate's very top module but not anything within its submodules. Meanwhile hovering & documentation pop-ups work for everything!!!
I believe this is related to the fact that the dependency has optional features.

## Detail
In the minimal example provided in this [repo](https://github.com/spamwax/rls-completion-features), you can see that external crate has a feature that needs to be enabled for unix platforms: `socket2 = { version = "*", features = ["unix"] }`

It seems that auto-completion for top-level modules/structs is working:

![ext_module](https://user-images.githubusercontent.com/1251233/62438066-c93d2500-b713-11e9-9af5-81bac13f1226.png)

but the inner `impl` or methods fails to appear in completion list:

![ext_module_methods](https://user-images.githubusercontent.com/1251233/62438095-ebcf3e00-b713-11e9-815c-2731ab8466b8.png)

What makes it more interesting is that
- hovering
- documentation
- Goto Definition
for those inner `impl` or methods **actually** work & shows their full definition and documentation, hinting that rls/racer is aware of them but for some reason can't show them in auto-completion:

![hover_method](https://user-images.githubusercontent.com/1251233/62438151-233dea80-b714-11e9-846a-011bb1b54350.png)

![ezgif-2-cd840908acb7](https://user-images.githubusercontent.com/1251233/62439595-79159100-b71a-11e9-9de2-c8d8290a2e7a.gif)

I have encountered similar issues in my other projects where either external crates or my own crate has some optional `features` that seems to confuse rls/racer.

The above test is on a fresh installation of vscode that only has Rust extension installed with following noteworthy settings:

```json
{
    "rust-client.channel": "stable",
    "rust.all_features": false,
    "rust.racer_completion": true,
    "rust.all_targets": true,
    "rust.features": [],
}
```
(I tried to play with some of those settings but didn't help!)

I am using

```
rls 1.36.0 (9692ca8 2019-05-18)
rustup 1.18.3 (435397f48 2019-05-22)
cargo 1.36.0 (c4fcfb725 2019-05-15)

(on x86_64 GNU/Linux)
```

Please let me know if more info is needed.

**Comment by chriskuech** [comment]
Is the rls-vscode team able to repro this issue?  This bug makes the extension unbeneficial for anyone trying to use Rust WASM, as the web-sys crate containing all the DOM bindings relies extensively on features.

**Comment by dmilford** [comment]
I'm also having troubles, and the web-sys crate is particularly difficult to use without intellisense.  I'm creating a tutorial video on 3D graphics in the browser, and it's quite embarrassing to not have intellisense working properly.  I have to know precisely what I'm typing, which is very difficult for some of the longer names.  

I would appreciate any hints as to how I can make this work correctly as the current situation really diminishes the value of Rust altogether.

**Comment by Sorrien** [comment]
I'm having the same problem. Is there any update on this?

**Comment by darman96** [comment]
Still no news/reaction? :(

**Comment by cleak** [comment]
My understanding is that the bug (likely multiple bugs) is either in RLS (https://github.com/rust-lang/rls) or in Racer (https://github.com/racer-rust/racer).

Specifically for WASM, RLS has an issue tracking it: https://github.com/rust-lang/rls/issues/1489

**Comment by chriskuech** [comment]
I haven't tried with `web-sys` specifically, but I have disabled the RLS extension and added the [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=matklad.rust-analyzer) extension in its place and it has been working much smoother so far.

**Comment by bergkvist** [comment]
### I'm using roxmltree=0.11.0 and I think I have the same problem.

I do sometimes get hints/docs about how a function within the library is used if I write the name (without being helped by intellisense along the way). I also get a red line below if I try to use something that doesn't exist. But no autocomplete.

**Comment by AntonBlomstrom** [comment]
> I haven't tried with `web-sys` specifically, but I have disabled the RLS extension and added the [rust-analyzer](https://marketplace.visualstudio.com/items?itemName=matklad.rust-analyzer) extension in its place and it has been working much smoother so far.

I just encountered this issue with rust-analyzer myself. 
![image](https://user-images.githubusercontent.com/39986991/81866951-42558c80-9570-11ea-8743-30113bbfef5a.png)

You can see that even annotating it as being a Window struct, it still gives an {unknown}
![image](https://user-images.githubusercontent.com/39986991/81867242-a9734100-9570-11ea-8f22-81ef424c7cf9.png)

**Comment by Xanewok** [comment]
(As a side note, you can set `"rust-client.engine": "rust-analyzer"` and use it directly from the same extension)

**Comment by JeremyAube** [comment]
> (As a side note, you can set `"rust-client.engine": "rust-analyzer"` and use it directly from the same extension)

I had the same problem, after setting `rust-client.engine` I restarted vs code and I got a prompt to install the nightly build of rust-analyzer. Everything works as expected now!

**Comment by bmhatfield** [comment]
Working through the rust book right now, and VSCode wasn't autocompleting `gen_range` in the [Guessing Game](https://doc.rust-lang.org/book/ch02-00-guessing-game-tutorial.html) tutorial. Switching to `rust-analyzer` as above fixes autocomplete for me. As a new user, it is _very_ confusing if the functions used in the tutorial in the book don't autocomplete.

**Comment by johnshew** [comment]
VSCode wasn't autocompleting gen_range in the Guessing Game tutorial. Switching to rust-analyzer as above fixes autocomplete for me.  Also had to set the auto update before it installed.

**Comment by visuallization** [comment]
Thanks for the hint. :) 
I have experienced an issue though.

Switching to `rust-analyzer` works for me and auto-completes e.g. `gen_range` but now warnings and errors will only be displayed in VSCode after manually building the project. It works automatically with the standard engine (which is rls, I guess). Any ideas how to fix this?

**Comment by lnicola** [comment]
@visuallization it's a known limitation of rust-analyzer, but you don't have to build the project manually since it runs cargo check on save. Moreso, you can use the auto-save feature of Code.

**Comment by mleonhard** [comment]
If the rust-analyzer download doesn't work, open VSCode Settings and set `Rust-analyzer: Release Tag` to `2020-08-31`.  Note that `stable` and `beta` don't work.  See #847.
